flang/lib/Lower/Bridge.cpp

```diff
@@ -1130,36 +1136,53 @@ class FirConverter : public Fortran::lower::AbstractConverter {
     builder->restoreInsertionPoint(insertPt);
   }
 
+  /// Generate FIR for a SELECT CASE statement.
+  /// The type may be CHARACTER, INTEGER, or LOGICAL.
   void genFIR(const Fortran::parser::SelectCaseStmt &stmt) {
     auto &eval = getEval();
     auto *context = builder->getContext();
     auto loc = toLocation();
     Fortran::lower::StatementContext stmtCtx;
     const auto *expr = Fortran::semantics::GetExpr(
         std::get<Fortran::parser::Scalar<Fortran::parser::Expr>>(stmt.t));
-    auto exprType = expr->GetType();
-    mlir::Value selectExpr;
-    if (isCharacterCategory(exprType->category())) {
-      TODO(loc, "Select Case selector of type Character");
+    bool isCharSelector = isCharacterCategory(expr->GetType()->category());
+    bool isLogicalSelector = isLogicalCategory(expr->GetType()->category());
```

Comment:
I think we have moved to using the FIR type for these tests. @jeanPerier ?

Reply:
In general yes, when there is a `mlir::Type` available working with it is the preferred way. But Here I think it would make the code more complex given this is needed before producing any mlir values, and working with `mlir::Type` would not be 100% straightforward given logicals are `mlir::IntegerType`. So what Val has here looks clean to me.

Reply:
Right, that was the motivation for doing it this way.